Show correct email's scheduled time when server and client are in different timezones [MAILPOET-5495]

NeosinneR opened the pull request:
## Description

This PR fixes the issue of showing the correct scheduled time for sending an email when the server and client time zones are different. When the email is scheduled, the date time in the database is saved without a time zone, practically meaning it's saved with the server's time zone in mind. However, on the front end, when the time zone is not provided, it assumes browser's time zone. 

Because of this, on the email listing page, the scheduled time is incorrectly rendered in the server timezone, which is misleading, and it also causes the UI to show `Not sent yet` for the duration of the time zone difference before the email is actually sent. 

**Example #1** - server time is 8:30 am, client time is 10:30 am. I schedule email to be sent in 15 minutes (8:45 am server, 10:45 am client).

<img width="435" alt="Screenshot 2024-06-11 at 10 30 47" src="https://github.com/mailpoet/mailpoet/assets/470616/4a47e004-803d-42a9-bf15-0fb0deef0e4d">

**Before this PR**, the client already thought that the email is past scheduled time, and because of that it showed `Not sent yet` message.

<img width="807" alt="Screenshot 2024-06-11 at 10 28 53" src="https://github.com/mailpoet/mailpoet/assets/470616/8ed27c9c-e83c-4cc8-ad04-99b2d1dc9e08">

**After this PR**, it correctly shows 10:45 am. 

<img width="760" alt="Screenshot 2024-06-11 at 10 29 55" src="https://github.com/mailpoet/mailpoet/assets/470616/7c5d9c8e-4e21-488d-b073-536b07dca766">

**Example #2** - server time is 8:30 am, client time is 10:30 am. I schedule email to be sent in 3 hours (11:30 am server, 1:30 pm client).

<img width="381" alt="Screenshot 2024-06-11 at 10 32 34" src="https://github.com/mailpoet/mailpoet/assets/470616/51eb8aa2-d9b0-45b9-b314-8ad1bf7793db">

**Before this PR**, the email listing page showed that the email would be sent at 11:30 am, which was misleading because it would be in fact sent at 1:30 pm client's time.

<img width="751" alt="Screenshot 2024-06-11 at 10 33 10" src="https://github.com/mailpoet/mailpoet/assets/470616/a618ed42-047c-46f2-8863-07e293ae94ca">

**After this PR**, it correctly shows 1:30 pm. 

<img width="734" alt="Screenshot 2024-06-11 at 10 32 21" src="https://github.com/mailpoet/mailpoet/assets/470616/f12377c8-9012-4ac2-a9d5-902a5f555edd">

## Code review notes

I haven't figured out how to write tests for this, as manipulating time zones on the front end is not very straightforward. Open to any suggestions.

## QA notes

You don't necessarily have to change the server time zone to test this, all it takes is to change the time zone of your computer. Please test using the same timezone, but also time zones that are ahead and behind the server (if the server is at +00:00, test in +00:00, +02:00, and -02:00). Also test scheduling the email within the timezone difference and outside of it (i.e., in the next 15 minutes, but also on the next day). 

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5495]

## After-merge notes

_N/A_

## Tasks

- [x] 🚫 I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [x] 🚫 I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5495]: https://mailpoet.atlassian.net/browse/MAILPOET-5495?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

NeosinneR committed:
Calculate server timezone offset in minutes

[MAILPOET-5495]

NeosinneR committed:
Create function to adjust for timezone difference between server and the client

[MAILPOET-5495]

NeosinneR committed:
Adjust newsletter sending time with timezone difference between server and client on emails listing page

[MAILPOET-5495]

triple0t commented on mailpoet/lib/AdminPages/Pages/Newsletters.php:
```diff
@@ -109,6 +109,7 @@ public function render() {
       ->format(DateTime::DEFAULT_DATE_FORMAT);
     $data['current_time'] = $dateTime->getCurrentTime();
     $data['current_date_time'] = $dateTime->getCurrentDateTime()->format(DateTime::DEFAULT_DATE_TIME_FORMAT);
+    $data['server_timezone_in_minutes'] = $dateTime->getServerTimezoneInMinutes();
```
Because `MailPoet.Date.adjustForTimezoneDifference()` can be used anywhere in the application, I would suggest we move this variable to a global place e.g [layout](https://github.com/mailpoet/mailpoet/blob/6e66bf1f4b77db32ae6fe260e1842303925f7d7d/mailpoet/views/layout.html) or anywhere else that works.

NeosinneR replied:
@triple0t nice suggestion 😉 Fixed in https://github.com/mailpoet/mailpoet/pull/5650/commits/c8afc972a25316deea1cb330263c635f22de68bd

triple0t reviewed (commented):
Looks good. 
I added a tiny comment for review

NeosinneR committed:
Expose server_timezone_in_minutes on all MailPoet pages, not just Emails

[MAILPOET-5495]

triple0t reviewed (dismissed):
LGTM

Aschepikov commented:
Hi @NeosinneR as I understand everything is ok on the send page, but on the list page the time is changing according **to the chosen UTC AND +2 as my real-time**, that's why I see some confusing things

UTC -8 shows my time -8 +2 on the email list page
![image](https://github.com/mailpoet/mailpoet/assets/113185437/5481d2c8-9f73-46dc-af34-ee4bf5e9ba22)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/d112eba4-03fa-4b2d-b22a-de78c3afd4b4)

UTC -2 shows not sent yet my current time -2 +2 looks like it is not in the past but I don't see the time
![image](https://github.com/mailpoet/mailpoet/assets/113185437/42014c3f-bd04-4ea6-866b-8c536d8e700a)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/dd3b71dc-d53f-45a1-8d54-5d415e420c76)


UTC +2  11:45 +2 +2 is 15:45 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/487ce947-9f81-4ea4-b767-821af9930768)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/62cadd54-20f1-4abf-8a36-9a6d9334b21a)

UTC +8 and with this time everything is tomorrow because +8 +2 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/b53638c8-9534-4710-a6d7-908d65ff7d1e)
On the email list, I see that it is tomorrow and there is no difference in what time I am choosing
![image](https://github.com/mailpoet/mailpoet/assets/113185437/9c590c97-f3c7-40b4-9bb7-06ff1cf9752a)

Aschepikov commented:
With UTC 0 it works correctly, but I am not sure if is it a desired behavior or not when there is another UTC in the settings

NeosinneR committed:
Fix calculating time zone difference and add comments with explanations

[MAILPOET-5495]

NeosinneR commented:
@Aschepikov nice catch! Since I've been operating in server +00:00 timezone, I missed this issue. The thing is, that PHP calculates time zone offset with a positive number when it's ahead of UTC, while JS calculates it as negative when it's ahead of UTC.

E.g., UTC-2 in PHP is -120 minutes, but it's +120 minutes in JS. 

I fixed this issue in https://github.com/mailpoet/mailpoet/commit/f3442c441f76041eed830e5ce1e5ab5fcebb650b and also added some commentary in the code.

Aschepikov reviewed (approved):
GJ!
